rust/connlib/libs/tunnel/src/tun_darwin.rs

```diff
@@ -115,54 +95,64 @@ impl IfaceDevice {
-        };
-
+    // On Apple platforms, we must use a NetworkExtension for reading and writing
+    // packets if we want to be allowed in the iOS and macOS App Stores. This has the
+    // unfortunate side effect that we're not allowed to create or destroy the tunnel
+    // interface ourselves. The file descriptor should already be opened by the NetworkExtension for us
+    // by this point. So instead, we iterate through the 1024 possible file descriptors
+    // looking for the one corresponding to the utun interface we have access to read and
+    // write from.
+    //
+    // Inspired heavily by WireGuard's implementation for Apple:
+    // https://github.com/WireGuard/wireguard-apple/blob/master/Sources/WireGuardKit/WireGuardAdapter.swift
+    pub async fn new() -> Result<Self> {
         let mut info = ctl_info {
             ctl_id: 0,
             ctl_name: [0; 96],
         };
-        info.ctl_name[..CTRL_NAME.len()]
+        info.ctl_name[..CTL_NAME.len()]
             // SAFETY: We only care about maintaining the same byte value not the same value,
             // meaning that the slice &[u8] here is just a blob of bytes for us, we need this conversion
             // just because `c_char` is i8 (for some reason).
             // One thing I don't like about this is that `ctl_name` is actually a nul-terminated string,
             // which we are only getting because `CTRL_NAME` is less than 96 bytes long and we are 0-value
             // initializing the array we should be using a CStr to be explicit... but this is slightly easier.
-            .copy_from_slice(unsafe { &*(CTRL_NAME as *const [u8] as *const [i8]) });
-
-        if unsafe { ioctl(fd, CTLIOCGINFO, &mut info as *mut ctl_info) } < 0 {
-            unsafe { close(fd) };
-            return Err(get_last_error());
-        }
-
-        let addr = sockaddr_ctl {
-            sc_len: size_of::<sockaddr_ctl>() as u8,
-            sc_family: AF_SYSTEM as u8,
-            ss_sysaddr: AF_SYS_CONTROL as u16,
-            sc_id: info.ctl_id,
-            sc_unit: idx,
-            sc_reserved: Default::default(),
-        };
-
-        if unsafe {
-            connect(
-                fd,
-                &addr as *const sockaddr_ctl as _,
-                size_of_val(&addr) as _,
-            )
-        } < 0
-        {
-            unsafe { close(fd) };
-            return Err(get_last_error());
+            .copy_from_slice(unsafe { &*(CTL_NAME as *const [u8] as *const [i8]) });
+
+        for fd in 0..1024 {
+            // initialize empty sockaddr_ctl to be populated by getpeername
+            let mut addr = sockaddr_ctl {
+                sc_len: size_of::<sockaddr_ctl>() as u8,
+                sc_family: 0 as u8,
+                ss_sysaddr: 0 as u16,
+                sc_id: info.ctl_id,
+                sc_unit: 0 as u32,
+                sc_reserved: Default::default(),
+            };
+
+            if unsafe {
+                getpeername(
```

Comment:
we should store the result in a variable and just check like:
`if ret != 0 || addr.sc_family != AF_SYSTEM`